// @codaworks/react-glow/dist/react-glow.cjs.development.js
"use client";
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];
      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }
    return target;
  };
  return _extends.apply(this, arguments);
}
function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;
  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }
  return target;
}

var _excluded = ["className", "size"],
  _excluded2 = ["className", "style", "children", "color", "debug"];
var GlowCapture = function GlowCapture(_ref) {
  var _ref$className = _ref.className,
    className = _ref$className === void 0 ? '' : _ref$className,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 400 : _ref$size,
    rest = _objectWithoutPropertiesLoose(_ref, _excluded);
  var element = React.useRef(null);
  React.useEffect(function () {
    var _element$current5, _element$current6;
    var move = function move(e) {
      if (e.pointerType === 'mouse') {
        requestAnimationFrame(function () {
          var _element$current, _element$current2;
          // @ts-ignore
          (_element$current = element.current) == null || _element$current.style.setProperty('--glow-x', e.layerX + "px");
          // @ts-ignore
          (_element$current2 = element.current) == null || _element$current2.style.setProperty('--glow-y', e.layerY + "px");
        });
      }
    };
    var leave = function leave() {
      var _element$current3, _element$current4;
      (_element$current3 = element.current) == null || _element$current3.style.removeProperty('--glow-x');
      (_element$current4 = element.current) == null || _element$current4.style.removeProperty('--glow-y');
    };
    (_element$current5 = element.current) == null || _element$current5.addEventListener('pointermove', move, {
      passive: true
    });
    (_element$current6 = element.current) == null || _element$current6.addEventListener('pointerleave', leave, {
      passive: true
    });
    return function () {
      var _element$current7, _element$current8;
      (_element$current7 = element.current) == null || _element$current7.removeEventListener('pointermove', move);
      (_element$current8 = element.current) == null || _element$current8.removeEventListener('pointerleave', leave);
    };
  }, []);
  return React__default.createElement("div", Object.assign({
    ref: element,
    className: "glow-capture " + className,
    style: {
      position: 'relative',
      '--glow-size': size + "px"
    }
  }, rest));
};
var mask = "\nradial-gradient(var(--glow-size) var(--glow-size) at calc(var(--glow-x, -999px) - var(--glow-left, 0px))\ncalc(var(--glow-y, -999px) - var(--glow-top, 0px)), #000000 1%, transparent 50%)\n";
var Glow = function Glow(_ref2) {
  var className = _ref2.className,
    style = _ref2.style,
    children = _ref2.children,
    _ref2$color = _ref2.color,
    color = _ref2$color === void 0 ? '#f50057' : _ref2$color,
    debug = _ref2.debug,
    rest = _objectWithoutPropertiesLoose(_ref2, _excluded2);
  var element = React.useRef(null);
  React.useEffect(function () {
    var _element$current9, _element$current10, _element$current11, _element$current12;
    (_element$current9 = element.current) == null || _element$current9.style.setProperty('--glow-top', ((_element$current10 = element.current) == null ? void 0 : _element$current10.offsetTop) + "px");
    (_element$current11 = element.current) == null || _element$current11.style.setProperty('--glow-left', ((_element$current12 = element.current) == null ? void 0 : _element$current12.offsetLeft) + "px");
  });
  React.useEffect(function () {
    var _element$current17;
    var observer = new ResizeObserver(function () {
      requestAnimationFrame(function () {
        var _element$current13, _element$current14, _element$current15, _element$current16;
        (_element$current13 = element.current) == null || _element$current13.style.setProperty('--glow-top', ((_element$current14 = element.current) == null ? void 0 : _element$current14.offsetTop) + "px");
        (_element$current15 = element.current) == null || _element$current15.style.setProperty('--glow-left', ((_element$current16 = element.current) == null ? void 0 : _element$current16.offsetLeft) + "px");
      });
    });
    var capture = (_element$current17 = element.current) == null ? void 0 : _element$current17.closest('.glow-capture');
    if (capture) observer.observe(capture);
    return function () {
      return observer.disconnect();
    };
  }, []);
  return React__default.createElement("div", Object.assign({
    ref: element,
    className: 'glow',
    style: {
      display: 'grid'
    }
  }, rest), React__default.createElement("div", Object.assign({
    className: className,
    style: _extends({}, style, {
      gridArea: '1/1/1/1'
    })
  }, rest), children), React__default.createElement("div", Object.assign({
    className: "glow-mask " + (className ? className : ''),
    glow: 'true',
    style: _extends({}, style, {
      '--glow-color': color,
      gridArea: '1/1/1/1',
      pointerEvents: 'none',
      mask: debug ? undefined : mask,
      WebkitMask: debug ? undefined : mask
    })
  }, rest), children));
};

exports.Glow = Glow;
exports.GlowCapture = GlowCapture;
